pufferlib/ocean/artillery3d/artillery3d.h

```diff
@@ -0,0 +1,687 @@
+            }
+        } else if (action == FUSETUP) {
+            if (env->fuse_t < 0.99) {
+                env->fuse_t += 0.01;
+            }
+            else {
+                env->score += env->out_bounds_penalty;
+            }
+        } else if (action == FUSETDOWN) {
+            if (env->fuse_t > 0.01) {
+                env->fuse_t -= 0.01;
+            }
+            else {
+                env->score += env->out_bounds_penalty;
+            }
+        }
+
+        if (action != FIRE) {
+            env->turn_penaltyn = get_turn_penalty(env);
+            if (env->turn_penaltyn > env->turn_penalty) env->turn_penaltyn = env->turn_penalty;
+            env->score += env->turn_penaltyn;
+            env->rewards[0] += env->turn_penaltyn;
+        }
+    }
+    else { // Projectile Active
+        env->projectile_time += TIMESTEP;
+        float pt = env->projectile_time;
+        env->px = X0 + env->vx0 * pt;
+        env->py = Y0 + env->vy0 * pt;
+        env->pz = Z0 + env->vz0 * pt - 0.5f * G * pt * pt;
+        if (env->px > XSIZE) {
+            env->fired = 0;
+            env->projectile_active = 0;
+        }
+        for (int i = 0; i < NUMTARGETS; i++) {
+            if (env->t > env->time_target_vanish[i]) {
+                env->tx[i] = 0.0f;
+                env->ty[i] = 0.0f;
+                env->tz[i] = 0.0f;
+                env->target_vx[i] = 0.0f;
+                env->target_vy[i] = 0.0f;
+                env->target_vz[i] = 0.0f;
+            }
+        }
+    }
+
+    for (int i = 0; i < NUMTARGETS; i++) {
+        if (env->tx[i] > 1.0f) {
+            env->tx[i] += env->target_vx[i] * TIMESTEP;
+            if (env->tx[i] < 0) env->tx[i] = 0.0f;
+            if (env->tx[i] > XSIZE) env->tx[i] = XSIZE;
+
+            env->ty[i] += env->target_vy[i] * TIMESTEP;
+            if (env->ty[i] < 0) {
+                env->ty[i] = 0.0f;
+                env->target_vy[i] = env->target_vy[i] * -1.0f;
+            }
+            if (env->ty[i] > YSIZE) env->ty[i] = YSIZE;
+
+            env->tz[i] += env->target_vz[i] * TIMESTEP;
+            if (env->tz[i] < 0) env->tz[i] = 0.0f;
+            if (env->tz[i] > ZSIZE) env->tz[i] = ZSIZE;
+        }
+    }
+
+    if (env->debug > 1) printf("  fired = %d remaining = %d\n", env->shots_fired, env->shots_remaining);
+    if (
+            (env->shots_remaining < 1 && (!env->render || env->px > XSIZE)) ||
+            (env->score < -1.0f) ||
+            (env->targets_remaining < 1 && (!env->render || env->px > XSIZE))
+        ) {
+        if (env->debug > 0) printf("==================terminate=================\n\n\n\n\n\n\n\n\n\n");
+        env->terminals[0] = 1;
+        add_log(env);
+        c_reset(env);
+    }
+    env->t += TIMESTEP;
+}
+
+void c_step(Artillery3D* env) {
+    env->terminals[0] = 0;
+    env->rewards[0] = 0.0;
+
+    int action = env->actions[0];
+    env->tick += 1;
+    step_frame(env, action);
+    compute_observations(env);
```

Comment:
Pretty solid env. Like plat. Few small things could be simplified here and there. Overall gestalt is good. Didn't catch any major messups on RL side. Having targets be position dependent is going to make learning a bit harder vs. sorting them. Not huge. Only other thing is possible the env is ticking too fast. Like atari with no frameskip. Nothing wrong with that if you want the benchmark to be a bit artificially harder. But you can solve the same problem easier on lower tick rate